### PR TITLE
gh-127681: add overflow checks when using  `PyMem_Malloc`

### PR DESCRIPTION
**Pull request by picnixz**
For now, this only adds checks in `Modules` and `Objects`.

* Issue: gh-127681

### CODE REVIEW

**Comment by picnixz** [comment]
I've added comments for future readers so that they know that something does not need to be checked or not, but I can also remove them if you think it's overly verbose.

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -331,19 +331,30 @@ _ctypes_alloc_format_string(const char *prefix, const char *suffix)
...
+    if (prefix) {
+        size_t prefix_len = strlen(prefix);
+        if (len > PY_SIZE_MAX - prefix_len) {
+            goto oom;
+        }
+        len += prefix_len;
+    }
+    if (len == PY_SIZE_MAX) {
+        goto oom;
+    }
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:343]: Do you have an example of overflow? AFAIK, `prefix` is smaller than 1030 bytes, and `suffix` also is a short string.

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -363,12 +374,16 @@ _ctypes_alloc_format_string_with_shape(int ndim, const Py_ssize_t *shape,
...
     prefix_len = 32 * ndim + 3;
-    if (prefix)
-        prefix_len += strlen(prefix);
+    if (prefix) {
+        size_t l = strlen(prefix);
+        if ((size_t)prefix > (size_t)PY_SSIZE_T_MAX - l) {
+            goto oom;
+        }
+        prefix_len += l;
+    }
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:383]: `prefix` is always "&" here.

File: Modules/_ctypes/_ctypes.c [hunk 3]
```diff
@@ -1666,7 +1685,7 @@ PyCArrayType_init(PyObject *self, PyObject *args, PyObject *kwds)
     if (stginfo->format == NULL)
         goto error;
     stginfo->ndim = iteminfo->ndim + 1;
-    stginfo->shape = PyMem_Malloc(sizeof(Py_ssize_t) * stginfo->ndim);
+    stginfo->shape = PyMem_New(Py_ssize_t, stginfo->ndim);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:1688]: This never overflows. `ndim < 32`.

File: Modules/_ctypes/_ctypes.c [hunk 4]
```diff
@@ -3094,7 +3113,7 @@ PyCData_MallocBuffer(CDataObject *obj, StgInfo *info)
         /* In python 2.4, and ctypes 0.9.6, the malloc call took about
            33% of the creation time for c_int().
         */
-        obj->b_ptr = (char *)PyMem_Malloc(info->size);
+        obj->b_ptr = PyMem_New(char, info->size);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:3116]: This is the same.

File: Modules/_ctypes/_ctypes.c [hunk 5]
```diff
@@ -4775,7 +4794,7 @@ Array_subscript(PyObject *myself, PyObject *item)
                 return PyBytes_FromStringAndSize(ptr + start,
                                                  slicelen);
             }
-            dest = (char *)PyMem_Malloc(slicelen);
+            dest = PyMem_New(char, slicelen);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:4797]: This is the same.

File: Modules/_ctypes/_ctypes.c [hunk 6]
```diff
@@ -5461,7 +5480,7 @@ Pointer_subscript(PyObject *myself, PyObject *item)
                 return PyBytes_FromStringAndSize(ptr + start,
                                                  len);
             }
-            dest = (char *)PyMem_Malloc(len);
+            dest = PyMem_New(char, len);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:5483]: This is the same.

File: Modules/_ctypes/stgdict.c [hunk 7]
```diff
@@ -43,35 +41,49 @@ PyCStgInfo_clone(StgInfo *dst_info, StgInfo *src_info)
     Py_XINCREF(dst_info->module);
 
     if (src_info->format) {
-        dst_info->format = PyMem_Malloc(strlen(src_info->format) + 1);
+        size_t s = strlen(src_info->format);
+        if (s >= (size_t)PY_SSIZE_T_MAX) {
+            goto oom;
+        }
+        dst_info->format = PyMem_Malloc(s + 1);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/stgdict.c:48]: This never overflows. `src_info->format` points to a memory array at least `strlen(src_info->format) + 1` bytes long.

File: Modules/_ctypes/stgdict.c [hunk 8]
```diff
@@ -43,35 +41,49 @@ PyCStgInfo_clone(StgInfo *dst_info, StgInfo *src_info)
...
         if (dst_info->format == NULL) {
-            PyErr_NoMemory();
-            return -1;
+            goto oom;
         }
         strcpy(dst_info->format, src_info->format);
     }
     if (src_info->shape) {
-        dst_info->shape = PyMem_Malloc(sizeof(Py_ssize_t) * src_info->ndim);
+        dst_info->shape = PyMem_New(Py_ssize_t, src_info->ndim);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/stgdict.c:55]: It never overflows. `ndim < 32`.

File: Modules/_ctypes/stgdict.c [hunk 9]
```diff
@@ -43,35 +41,49 @@ PyCStgInfo_clone(StgInfo *dst_info, StgInfo *src_info)
...
                sizeof(Py_ssize_t) * src_info->ndim);
     }
 
     if (src_info->ffi_type_pointer.elements == NULL)
         return 0;
-    size = sizeof(ffi_type *) * (src_info->length + 1);
+    if ((size_t)src_info->length > (size_t)PY_SSIZE_T_MAX / sizeof(ffi_type *) - 1) {
+        goto oom;
+    }
+    Py_ssize_t size = sizeof(ffi_type *) * (src_info->length + 1);
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/stgdict.c:68]: Do you have an example of overflow?

File: Modules/_ctypes/stgdict.c [hunk 10]
```diff
@@ -331,6 +343,10 @@ PyCStructUnionType_update_stginfo(PyObject *type, PyObject *fields, int isStruct
         PyMem_Free(stginfo->format);
         stginfo->format = NULL;
     }
+    if (format_spec_size >= PY_SSIZE_T_MAX) {
+        PyErr_NoMemory();
+        goto error;
+    }
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/stgdict.c:349]: It never overflows. There is a memory array at least `format_spec_size+1` bytes long.

File: Modules/_io/textio.c [hunk 11]
```diff
@@ -465,6 +465,10 @@ _PyIncrementalNewlineDecoder_decode(PyObject *myself,
                when there is something to translate. On the other hand,
                we already know there is a \r byte, so chances are high
                that something needs to be done. */
+            if (len > PY_SSIZE_T_MAX / kind) {
+                PyErr_NoMemory();
+                goto error;
+            }
```
Comment (serhiy-storchaka) [inline Modules/_io/textio.c:471]: It never overflows. The size of `output` is larger than `kind * len`.

File: Modules/_io/winconsoleio.c [hunk 12]
```diff
@@ -610,7 +610,7 @@ static wchar_t *
 read_console_w(HANDLE handle, DWORD maxlen, DWORD *readlen) {
     int err = 0, sig = 0;
 
-    wchar_t *buf = (wchar_t*)PyMem_Malloc(maxlen * sizeof(wchar_t));
+    wchar_t *buf = PyMem_New(wchar_t, maxlen);
```
Comment (serhiy-storchaka) [inline Modules/_io/winconsoleio.c:613]: It never overflows. `maxlen` is at most `DWORD_MAX / 4`, and `sizeof(wchar_t)` is 2.

File: Modules/_io/winconsoleio.c [hunk 13]
```diff
@@ -875,8 +875,10 @@ _io__WindowsConsoleIO_readall_impl(winconsoleio *self)
         return NULL;
 
     bufsize = BUFSIZ;
-
-    buf = (wchar_t*)PyMem_Malloc((bufsize + 1) * sizeof(wchar_t));
+    // Note: BUFSIZ is small enough not to overflow, so no need
+    //       to check that 'bufsize + 1' overflows. However, we
+    //       still need to check (bufsize + 1) * sizeof(wchar_t).
+    buf = PyMem_New(wchar_t, bufsize);
```
Comment (serhiy-storchaka) [inline Modules/_io/winconsoleio.c:881]: It never overflows.

File: Modules/_posixsubprocess.c [hunk 14]
```diff
@@ -1125,13 +1125,16 @@ subprocess_fork_exec_impl(PyObject *module, PyObject *process_args,
         if (extra_group_size < 0)
             goto cleanup;
 
-        if (extra_group_size > MAX_GROUPS) {
+        if (extra_group_size > MAX_GROUPS
+            || (size_t)extra_group_size > (size_t)PY_SSIZE_T_MAX / sizeof(gid_t))
```
Comment (serhiy-storchaka) [inline Modules/_posixsubprocess.c:1129]: We can be sure that MAX_GROUPS is small enough.

File: Modules/_posixsubprocess.c [hunk 15]
```diff
@@ -1192,7 +1195,7 @@ subprocess_fork_exec_impl(PyObject *module, PyObject *process_args,
 #endif /* HAVE_SETREUID */
     }
 
-    c_fds_to_keep = PyMem_Malloc(fds_to_keep_len * sizeof(int));
+    c_fds_to_keep = PyMem_New(int, fds_to_keep_len);
```
Comment (serhiy-storchaka) [inline Modules/_posixsubprocess.c:1198]: It never overflows. `fds_to_keep_len` is the size of a tuple. `sizeof(int) <= sizeof(PyObject *)`.

File: Modules/_struct.c [hunk 16]
```diff
@@ -1725,7 +1725,7 @@ prepare_s(PyStructObject *self)
     }
 
     /* check for overflow */
-    if ((ncodes + 1) > ((size_t)PY_SSIZE_T_MAX / sizeof(formatcode))) {
+    if (ncodes > ((size_t)PY_SSIZE_T_MAX / sizeof(formatcode)) - 1) {
```
Comment (serhiy-storchaka) [inline Modules/_struct.c:1728]: `ncodes + 1` does not overflow. `ncodes` is not longer then the length of str/bytes.

File: Objects/abstract.c [hunk 17]
```diff
@@ -705,8 +705,7 @@ int PyObject_CopyData(PyObject *dest, PyObject *src)
 
     /* Otherwise a more elaborate copy scheme is needed */
 
-    /* XXX(nnorwitz): need to check for overflow! */
-    indices = (Py_ssize_t *)PyMem_Malloc(sizeof(Py_ssize_t)*view_src.ndim);
+    indices = PyMem_New(Py_ssize_t, view_src.ndim);
```
Comment (serhiy-storchaka) [inline Objects/abstract.c:708]: It does not overflow. `ndim < 32`.

File: Objects/call.c [hunk 18]
```diff
@@ -484,10 +484,13 @@ _PyObject_Call_Prepend(PyThreadState *tstate, PyObject *callable,
     PyObject **stack;
 
     Py_ssize_t argcount = PyTuple_GET_SIZE(args);
-    if (argcount + 1 <= (Py_ssize_t)Py_ARRAY_LENGTH(small_stack)) {
+    if (argcount <= (Py_ssize_t)Py_ARRAY_LENGTH(small_stack) - 1) {
```
Comment (serhiy-storchaka) [inline Objects/call.c:487]: `argcount + 1` does not overflow. `argcount` is the length of a tuple.

File: Objects/listobject.c [hunk 19]
```diff
@@ -3637,8 +3637,7 @@ list_ass_subscript_lock_held(PyObject *_self, PyObject *item, PyObject *value)
                 step = -step;
             }
 
-            garbage = (PyObject**)
-                PyMem_Malloc(slicelength*sizeof(PyObject*));
+            garbage = PyMem_New(PyObject *, slicelength);
```
Comment (serhiy-storchaka) [inline Objects/listobject.c:3640]: It does not overflow. `slicelength` is the length of a slice in existing list. The list already uses at least `slicelength*sizeof(PyObject*)` bytes.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
I reviewed, and found only one site where an integer overflow _may_ be possible. In all other cases it is not possible.

**Comment by picnixz** [comment]
So actually, there are parts that could be removed instead of fortified. I won't work on this one today but my question is: can I actually add the comments so that we know that there won't be someone else wondering "oh maybe we should check for this"?

**Comment by serhiy-storchaka** [comment]
I don't like pollution of the code with trivial comments. They actually make reading the code more difficult -- you need to read and comprehend the comment instead of just skipping the presumably correct code. It is responsibility of the contributor to prove that they changes are necessary. The CPython code is mature enough, and most trivial bugs are fixed.

So I suggest you to close this PR and open a new PR with only cases for which you can show at least theoretical possibility of integer overflow.

**Pull request closed without merging**